// nativefier/lib/constants.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PLACEHOLDER_APP_DIR = exports.ELECTRON_MAJOR_VERSION = exports.DEFAULT_SAFARI_VERSION = exports.DEFAULT_FIREFOX_VERSION = exports.DEFAULT_CHROME_VERSION = exports.DEFAULT_ELECTRON_VERSION = exports.DEFAULT_APP_NAME = void 0;
const path = __importStar(require("path"));
exports.DEFAULT_APP_NAME = 'APP';
// Update both DEFAULT_ELECTRON_VERSION and DEFAULT_CHROME_VERSION together,
// and update app / package.json / devDeps / electron to value of DEFAULT_ELECTRON_VERSION
exports.DEFAULT_ELECTRON_VERSION = '13.4.0';
// https://atom.io/download/atom-shell/index.json
exports.DEFAULT_CHROME_VERSION = '91.0.4472.164';
// Update each of these periodically
// https://product-details.mozilla.org/1.0/firefox_versions.json
exports.DEFAULT_FIREFOX_VERSION = '92.0';
// https://en.wikipedia.org/wiki/Safari_version_history
exports.DEFAULT_SAFARI_VERSION = {
    majorVersion: 15,
    version: '15.0',
    webkitVersion: '605.1.15',
};
exports.ELECTRON_MAJOR_VERSION = parseInt(exports.DEFAULT_ELECTRON_VERSION.split('.')[0], 10);
exports.PLACEHOLDER_APP_DIR = path.join(__dirname, './../', 'app');
